include/modules/meta/factory.hpp

```diff
@@ -43,73 +42,71 @@
 #if ENABLE_XKEYBOARD
 #include "modules/xkeyboard.hpp"
 #endif
-#if not(ENABLE_I3 && ENABLE_MPD && ENABLE_NETWORK && ENABLE_ALSA && ENABLE_PULSEAUDIO && ENABLE_CURL && ENABLE_XKEYBOARD)
 #include "modules/unsupported.hpp"
-#endif
 
 POLYBAR_NS
 
 using namespace modules;
 
 namespace {
   module_interface* make_module(string&& name, const bar_settings& bar, string module_name, const logger& m_log) {
```

Comment:
(Not for this PR)
We should use a sort of global unordered_map and register each module in a map
This function is so ugly...

Reply:
I agree. I tried to make it a bit nicer in this PR by moving the type string into the module header but still.

But I'm not really sure how to achieve this, C++ doesn't have classes at runtime (only instances), so you can't do something like this: `clazz.newInstance()` as you can do in java.

Reply:
You can do something like:
```
struct BaseModuleFactory  {
    BaseModuleFactory() { /* do stuff */ }

    virtual ModuleInterface* build(...) = 0;
};

template <typename T>
struct ModuleFactory : BaseModuleFactory {
   ModuleInterface* build(...) { return new T(...); }
};

ModuleFactory<MyModule> var;
```

Reply:
Oh, and then we map module types to `ModuleFactory` instances?

Reply:
Yeap
```C++
// Factory header
void registerModuleFactory(std::string name, BaseModuleFactory& f);

// Factory source
static std::unordered_map<std::string, BaseModuleFactory&> getFactories(); // Returns factory map to avoid static order initialization fiasco
void registerModuleFactory(std::string name, BaseModuleFactory& f)  {
  getFactories()[name] = f;
}
```

Reply:
Neat! We should definitely look into that. But maybe after this PR is merged.